### PR TITLE
Allow users to see oldest messages first

### PR DESCRIPTION
**Pull request by ChristophWurst**
Fixes https://github.com/nextcloud/mail/issues/679

Tasks
* [x] Add UI to toggle between newest first or oldest first
* [x] Send search preference to back-end with each query
* [x] Adjust front-end data storage to search preference
* [x] Fix sync for new messages when latest page is reached
* [x] Adjust `needsFetch` comparisons to differently ordered timestamps

Follow-ups
* Possibly allow switching the sort order without a full page reload

### CODE REVIEW

File: lib/Service/Sync/SyncService.php [hunk 1]
```diff
@@ -151,24 +158,35 @@ public function syncMailbox(Account $account,
 	private function getDatabaseSyncChanges(Account $account,
 											Mailbox $mailbox,
 											array $knownIds,
+											?int $lastMessageTimestamp,
+											string $sortOrder,
 											?SearchQuery $query): Response {
-		if (empty($knownIds)) {
+		if (empty($knownIds) || $lastMessageTimestamp === null) {
```
Comment (ChristophWurst) [inline lib/Service/Sync/SyncService.php:164]:
```suggestion
		if (empty($knownIds) && $lastMessageTimestamp === null) {
```

but not sure

**Comment by kesselb** [comment]
Priority inbox / All inboxes

Screenshot from 2021-09-20 21-02-41

I have two mail accounts configured. On scroll down the next messages are loaded from both accounts. We see the oldest messages from mail account A and B next to each other.

File: src/store/actions.js [hunk 2]
```diff
@@ -94,7 +95,13 @@ const findIndividualMailboxes = curry((getMailboxes, specialRole) =>
 	)
 )
 
-const combineEnvelopeLists = pipe(flatten, orderBy(prop('dateInt'), 'desc'))
+const combineEnvelopeLists = (sortOrder) => {
+	if (sortOrder === 'oldest-first') {
+		return pipe(flatten, orderBy(prop('dateInt'), 'desc'))
```
Comment (kesselb) [inline src/store/actions.js:100]: This looks wrong but it seems to work locally with desc :man_shrugging:

File: src/store/mutations.js [hunk 3]
```diff
@@ -170,29 +174,53 @@ export default {
...
 			})
+			return
+		}
+
+		envelopes.forEach((envelope) => {
+			const existing = mailbox.envelopeLists[listId] || []
+			normalizeTags(state, envelope)
+			Vue.set(state.envelopes, envelope.databaseId, Object.assign({}, state.envelopes[envelope.databaseId] || {}, envelope))
+			Vue.set(envelope, 'accountId', mailbox.accountId)
+			Vue.set(mailbox.envelopeLists, listId, uniq(orderByDateInt(existing.concat([envelope.databaseId]))))
```
Comment (kesselb) [inline src/store/mutations.js:209]: Could we update the existing envelope list in one go for all envelopes?

**Comment by ChristophWurst** [comment]
> I have two mail accounts configured. On scroll down the next messages are loaded from both accounts. We see the oldest messages from mail account A and B next to each other.

My observation: the first page is correct, then, when you load more, it actually doesn't just add messages to the end of the list but also in between. This is why the previously last element also jumps.

File: src/store/actions.js (hunk 2 above)
Comment (ChristophWurst) [reply src/store/actions.js:100]: This is wrong and caused https://github.com/nextcloud/mail/pull/5495#issuecomment-929907448. After I changed to asc that quirk goes away. Thanks :hugs:

**Comment by ChristophWurst** [comment]
> I have two mail accounts configured. On scroll down the next messages are loaded from both accounts. We see the oldest messages from mail account A and B next to each other.

I can't reproduce this. Could you try with the latest version?

**Comment by GretaD** [comment]
I tested this one and i noticed that when i click show oldest first, the first active message its not open, so the third column is empty. Same when i change it to newest.

File: lib/Db/MessageMapper.php [hunk 4]
```diff
@@ -151,13 +153,32 @@ public function findAllUids(Mailbox $mailbox): array {
...
 
+		if ($limit !== null) {
+			$query->setMaxResults($limit);
+		}
+		if ($sortOrder === IMailSearch::ORDER_NEWEST_FIRST) {
+			$query->orderBy('sent_at', 'desc');
+		} elseif ($sortOrder === IMailSearch::ORDER_OLDEST_FIRST) {
+			$query->orderBy('sent_at', 'asc');
+		} elseif ($sortOrder !== null) {
+			throw new InvalidArgumentException('Invalid sort orderd');
```
Comment (jancborchardt) [inline lib/Db/MessageMapper.php:179]:
```suggestion
			throw new InvalidArgumentException('Invalid sort order');
```

**Comment by ChristophWurst** [comment]
Stale and terribly conflicting

**Pull request closed without merging**